app/services/report_importer.rb

```diff
@@ -44,6 +59,15 @@ def import
     end
   end
 
+  def scan
+    logger.info "Scanning report with: #{self.class.report_scanner.join(', ')}"
+    self.class.report_scanner.each do |scanner|
+      scanner.scan(report)
+      logger.debug "Changes after #{scanner}: #{report.changes.inspect}"
+    end
+    report.save
```

Comment:
this adds another save call to the importing chain, could it be removed and the scanning moved just after logs are available but report is not saved yet at https://github.com/bastilian/foreman/blob/9346bf19279d7085090e690f780d20216cbf4596/app/services/report_importer.rb#L160 ?

Reply:
Having it save separately and not within the process of creating the logs, ensures that in any case the logs and report are saved. Cases where the scanner could "corrupt" a report maybe rare, but they could.

Another advantage of including a save when scanning is that we can have them run independently as well.

I agree with trying to keep writes low, but also trying to asses if it is useful here.
Maybe having saving as an optional argument to pass would satisfy both.

Reply:
We could have it optional if you feel strongly it's useful, but I think it should default to one save per repotr. Think about 10 000 hosts running puppet agent every 30 minutes, that gives you 0.003s per one report, so I think we should keep this part as efficient as possible.

Reply:
Sorry typing too fast, 0.18s